backend/services/user_groups.go

```diff
@@ -0,0 +1,421 @@
+		return backend.MissingValueErr("Slug")
+	}
+	if i.Name == "" {
+		return backend.MissingValueErr("Name")
+	}
+	return nil
+}
+
+func AddUsersToGroup(tx *database.Transactable, userSlugs []string, groupID int64) error {
+	if len(userSlugs) > 0 {
+		questionMarks := "("
+
+		interfaceSlice := make([]interface{}, len(userSlugs)+1)
+		for i, v := range userSlugs {
+			questionMarks += "?, "
+			interfaceSlice[i] = v
+		}
+
+		questionMarks = strings.TrimSuffix(questionMarks, ", ")
+		questionMarks += ")"
+
+		sqlStatement := fmt.Sprintf(`INSERT IGNORE INTO group_user_map(user_id, group_id)
+					SELECT users.id, user_groups.id
+					FROM users, user_groups
+					WHERE users.slug in %s and user_groups.id = ?;`, questionMarks)
+
+		interfaceSlice[len(userSlugs)] = groupID
+		err := tx.Exec(sq.Expr(sqlStatement, interfaceSlice...))
+
+		if err != nil {
+			return backend.WrapError("Unable to add users to group", backend.DatabaseErr(err))
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func CreateUserGroup(ctx context.Context, db *database.Connection, i CreateUserGroupInput) (*dtos.UserGroup, error) {
+	if err := isAdmin(ctx); err != nil {
+		return nil, backend.WrapError("Unwilling to create a user group", backend.UnauthorizedReadErr(err))
+	}
+
+	cleanSlug := SanitizeSlug(i.Slug)
+	if cleanSlug == "" {
+		return nil, backend.BadInputErr(errors.New("Unable to create operation. Invalid operation slug"), "Slug must contain english letters or numbers")
+	}
+
+	err := db.WithTx(context.Background(), func(tx *database.Transactable) {
+		id, _ := tx.Insert("user_groups", map[string]interface{}{
+			"slug": cleanSlug,
+			"name": i.Name,
+		})
+		AddUsersToGroup(tx, i.UserSlugs, id)
+	})
+
+	if err != nil {
+		return nil, backend.WrapError("Error creating user group", backend.BadInputErr(err, "A user group with this slug already exists; please choose another name"))
+	}
+	return &dtos.UserGroup{
+		Slug: i.Slug,
+		Name: i.Name,
+	}, nil
+}
+
+func ModifyUserGroup(ctx context.Context, db *database.Connection, i ModifyUserGroupInput) (*dtos.UserGroup, error) {
+	if err := isAdmin(ctx); err != nil {
+		return nil, backend.WrapError("Unwilling to modify a user group", backend.UnauthorizedReadErr(err))
+	}
+
+	if err := i.validateUserGroupInput(); err != nil {
+		return nil, backend.WrapError("Unable to modify user group", backend.BadInputErr(err, "Unable to modify user group due to bad input"))
+	}
+
+	userGroup, err := lookupUserGroup(db, i.Slug)
+	if err != nil {
+		return nil, backend.WrapError("Unable to modify user group", backend.UnauthorizedWriteErr(err))
+	}
+
+	err = db.WithTx(context.Background(), func(tx *database.Transactable) {
+		if i.Name != "" {
+			tx.Update(sq.Update("user_groups").Set("name", i.Name).Where(sq.Eq{"id": userGroup.ID}))
+		}
+		if len(i.UsersToRemove) > 0 {
+			for _, userSlug := range i.UsersToRemove {
+				tx.Exec(sq.Expr("DELETE gm FROM group_user_map gm JOIN users u on gm.user_id = u.id WHERE u.slug=?;", userSlug))
+			}
+		}
```

Comment:
I think you might be able to replace the loop with:

```go
tx.Exec(sq.Expr("DELETE gm FROM group_user_map gm JOIN users  u  ON gm.user_id = u.id WHERE u.slug in(?)", i.UsersToRemove))
```
Squirrel seems to support lists as parameters, so it might be possible. I didn't check though.

Unfortunately, it looks like [This  PR](https://github.com/Masterminds/squirrel/pull/295) needs to be merged before you can use the delete query builder.

Reply:
Ah, hm, actually this might not work like I expect --  you might need to use  Sq.Eq, which would _probably_ mean that you'd have to use a deleteBuilder, which, per the PR above, won't work. :( You might be able to do something like you did in addUsers though -- joining the slugs and appending a dynamic number of elements.

Reply:
modified it to use the ?'s like I was previously doing in addUsers!